### PR TITLE
Refactor DC Woodbury

### PR DESCRIPTION
**Pull request by SylvestreSakti**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**
No

**What kind of change does this PR introduce?**
This PR follows this PR: https://github.com/powsybl/powsybl-open-loadflow/pull/1109
Some variable names and comments have been modified to ensure a better maintanability of the code

**What is the current behavior?**

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [x] No

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java [hunk 1]
```diff
@@ -496,13 +499,15 @@ public void analyse(Network network, List<PropagatedContingency> contingencies,
...
+    private static void matrixCopyValues(DenseMatrix originalMatrix, DenseMatrix copyMatrix) {
         if (originalMatrix.getRowCount() == copyMatrix.getRowCount() && originalMatrix.getColumnCount() == copyMatrix.getColumnCount()) {
             for (int columnIndex = 0; columnIndex < originalMatrix.getColumnCount(); columnIndex++) {
                 for (int rowIndex = 0; rowIndex < originalMatrix.getRowCount(); rowIndex++) {
                     copyMatrix.set(rowIndex, columnIndex, originalMatrix.get(rowIndex, columnIndex));
                 }
             }
+        } else {
+            throw new IllegalArgumentException();
         }
```
Comment (vidaldid-rte) [inline src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:511]: Maybe a message in the exception. For example. Invalid matrix size for copyMatrix. x columns and r rows expected. Received x' columns and r' rows.

File: src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java [hunk 2]
```diff
@@ -108,7 +108,7 @@ private double[] calculatePostContingencyStatesForAContingency(DcLoadFlowContext
             if (!disabledBuses.isEmpty() || !lostPhaseControllers.isEmpty()) {
                 newFlowStates = DcLoadFlowEngine.run(loadFlowContext, disabledNetwork, reportNode);
             }
-            postContingencyStates = engine.run(newFlowStates);
+            engine.computeToPostContingencyStates(newFlowStates);
         } else {
```
[the remark this answers is not in the document]
Comment (vidaldid-rte) [reply src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java:112]: In this file(WoodburyDCSecurityAnalysis), line 285, can you rename baseContingencyStates to workingContingencyStates to highlight that the array is reused for each contingency ?

**Comment by vidaldid-rte** [comment]
There are still type of contingenies in DcSensitivyAnalysis where a new DenseMatrix is created  (For example in caculateFactorStates. This can lead to the same memory issues in cases where many contingency fall in such category.

It would be good to have a resizableDenseMatrix that can grow its directBuffer if needed when resized, and could be used as a working matrix for all contingencies. 
In the same idea, for WoodburyDCSecurityAnalysis, we could have a large reusable array of double that could grow when needed.

File: src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java (hunk 1 above)
Comment (SylvestreSakti) [reply src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:512]: Yes, I modified this exception to a MatrixException (same check is done and same exception is used when summing two matrices in powybl-core)

**Comment by SylvestreSakti** [comment]
> There are still type of contingenies in DcSensitivyAnalysis where a new DenseMatrix is created (For example in caculateFactorStates. This can lead to the same memory issues in cases where many contingency fall in such category.
> 
> It would be good to have a resizableDenseMatrix that can grow its directBuffer if needed when resized, and could be used as a working matrix for all contingencies. In the same idea, for WoodburyDCSecurityAnalysis, we could have a large reusable array of double that could grow when needed.
> 
> But this is maybe out of the scope of this PR since it needs an evolution i core.

Yes, I agree with that, a "ResizableDenseMatrix" would help using less memory. But indeed may be out of this scope (and would need a modification in powsybl-core)

**Review by vidaldid-rte — APPROVED** [review]

File: src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java [hunk 3]
```diff
@@ -201,8 +201,10 @@ protected SecurityAnalysisResult runSimulations(LfNetwork lfNetwork, List<Propag
             // this is a difference with dc security analysis
             cleanContingencies(lfNetwork, propagatedContingencies);
 
+            // compute the pre-contingency states
             double[] preContingencyStates = DcLoadFlowEngine.run(context, new DisabledNetwork(), reportNode);
-            double[] baseContingencyStates = preContingencyStates.clone();
+            // create workingContingencyStates that will be a working copy of pre-contingency states
+            double[] workingContingencyStates = preContingencyStates.clone();
```
Comment (jeandemanged) [inline src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java:207]: Use `System.arraycopy` instead of clone ?

**Review by jeandemanged — CHANGES_REQUESTED** [review]

File: src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java (hunk 3 above)
Comment (SylvestreSakti) [reply src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java]: Done ! I did not know about this subtle difference but seems slightly better indeed (https://stackoverflow.com/questions/7179251/is-there-any-reason-to-prefer-system-arraycopy-over-clone)

Comment (vidaldid-rte) [reply src/main/java/com/powsybl/openloadflow/sa/WoodburyDcSecurityAnalysis.java]: That was true for Java5 in 2006 but if you run the bench today you see no difference after the JIT has comleted.

**Review by Hadrien-Godard — APPROVED** [review]

**Review by jeandemanged — APPROVED** [review]

**Pull request merged**